### ISSUE TITLE
builder crashes on kube-cluster after cluster VMs get rebooted 

### ISSUE THREAD

**Issue by rimusz** [deis/builder#379]
# Summary

v2.0.0 builder crashes on kube-cluster after cluster VMs get rebooted

## If You Are Reporting a Bug

After successful deis-workflow 2.0.0 install on [kube-cluster](https://github.com/TheNewNormal/kube-cluster-osx), builder there works just fine.

### Replication Steps

Please provide a detailed list of steps to reproduce it.

1. Boot kube-cluster
2. Install Workflow with 'install_deis' script, which install Workflow with the default minio support
3. check with `kd get po` and builder shows Running status.
4. Halt the Up via menu to start VMs again
5. 'kd get po` shows `deis-builder-kpd52            0/1       CrashLoopBackOff   18         17h`
6. `kd logs deis-builder-kpd52` shows
```
2016/07/01 11:48:13 Error getting storage parameters (read /var/run/secrets/deis/objectstore/creds/..7987_01_07_11_42_31.882873540: is a directory)
```

Running Workflow on kube-solo, there is no such problem with the builder

**Comment by rimusz** [comment]
not the issue anymore on Workflow v2.1, closing this now down

**Comment by rimusz** [comment]
operantly the problem came back on kubernetes v1.3

**Comment by brunoalano** [comment]
I'm with the same issue

**Comment by bacongobbler** [comment]
Sounds to me like an upstream issue with secrets on a rebooted node. From judging https://github.com/kubernetes/kubernetes/issues/7351 it doesn't look like they have a valid strategy for rebooted nodes. Try destroying it and booting a new node instead.

**Comment by bacongobbler** [comment]
I stand corrected; while this does seem odd that there is a stale file in the volume, we should have smarter logic [here](https://github.com/deis/builder/blob/master/pkg/conf/config.go#L48-L64) to determine what files we are looking for.

**Comment by jackfrancis** [comment]
I am still able to reproduce a related error using k8s `1.3.0` and workflow `v2.1.0`, it appears there are files that pass the "is not a directory" test, but still foil builder. It's possible we want to check additionally for "regular" files (`os.FileInfo.Mode().IsRegular()`) where we are checking to make sure we're not treating a directory like a file.

**Comment by felixbuenemann** [comment]
@jackfrancis I don't think `IsRegular()` will help, because the file we want to read are also symlinks, but pointing to files in `..data` which is a symlink to the actual directory.

The easiest is probably to just skip, if the filename is `..data`. A more complicated approach would be to check for a symlink, resolve it, skip if is's a dir.

**Comment by jackfrancis** [comment]
@felixbuenemann I think adding code to skip `..data` will prove very non-resilient pretty quickly, so if the generic failure condition is symlink-->dir, then the best thing would be to account for that condition, and skip as you outlined. This way when another `..data`-like thing emerges, we're resilient against it.

**Comment by bacongobbler** [comment]
I'd be happy to merge this as-is to get this issue fixed for v2.2, but I agree that we should look at more resilient solutions in the future like what you suggested, @jackfrancis.

**Comment by felixbuenemann** [comment]
@jackfrancis We can always add a more sophisticated approach in the future, if kubernetes changes their code to require it.

Doing that right now would probably end up with a solution that might or might not be future proof.
For example they could use symlinks to symlinks, which would then require recursion…
